evalml/pipelines/component_graph.py

```diff
@@ -0,0 +1,372 @@
+
+        Arguments:
+            X (pd.DataFrame): The input training data of shape [n_samples, n_features]
+            y (pd.Series): The target training data of length [n_samples]
+        """
+        return self._compute_features(self.compute_order[:-1], X, y, fit=True)
+
+    def predict(self, X):
+        """Make predictions using selected features.
+
+        Arguments:
+            X (pd.DataFrame): Data of shape [n_samples, n_features]
+
+        Returns:
+            pd.Series: Predicted values.
+        """
+        return self._compute_features(self.compute_order, X)
+
+    def compute_final_component_features(self, X, y=None):
+        """ Transform all components save the final one, and gathers the data from any number of parents
+        to get all the information that should be fed to the final component
+
+        Arguments:
+            X (pd.DataFrame): Data of shape [n_samples, n_features]
+
+        Returns:
+            pd.DataFrame: Transformed values.
+        """
+        if len(self.compute_order) <= 1:
+            return X
+        x_inputs = []
+        for parent in self.get_parents(self.compute_order[-1]):
+            if parent[-2:] == '.x' or parent[-2] == '.y':
+                parent = parent[:-2]
+            parent_path = self._get_ancestors(parent)
+            parent_path.append(parent)
+            parent_input = self._compute_features(parent_path, X, y=y)
+            if isinstance(parent_input, pd.Series):
+                parent_input = pd.DataFrame(parent_input, columns=[parent])
+            x_inputs.append(parent_input)
+        X_t = pd.concat(x_inputs, axis=1)
+        return X_t
+
+    def _get_ancestors(self, component):
+        ancestors = []
+        parents = self.get_parents(component)
+        while len(parents) > 0:
+            parent = parents.pop()
+            if parent[-2:] == '.x' or parent[-2:] == '.y':
+                parent = parent[:-2]
+            ancestors.append(parent)
+            parents.extend(self.get_parents(parent))
+        ancestors.reverse()
+        return ancestors
+
+    def _compute_features(self, component_list, X, y=None, fit=False):
+        """Transforms the data by applying the given components.
+
+        Arguments:
+            component_list (list): The list of component names to compute.
+            X (pd.DataFrame): Input data to the pipeline to transform.
+            y (pd.Series): The target training data of length [n_samples]
+            fit (bool): Whether to fit the estimators as well as transform it.
+                        Defaults to False.
+
+        Returns:
+            pd.DataFrame or pd.Series - Output of the last component
+        """
+        if not isinstance(X, pd.DataFrame):
+            X = pd.DataFrame(X)
+
+        output_cache = {}
+        final_component = None
+        for component_name in component_list:
+            final_component = component_name
+            component_class = self.get_component(component_name)
```

Comment:
Nit-pick: could we call this `component` or `component_instance` instead of `component_class`? Because its an instance, not a class.